### PR TITLE
fix(listbox+combobox): render options panel outside of scrollable container AL-4547

### PR DESCRIPTION
**Pull request by manuatarchilogic**
Fix feels a bit dirty, I don't like to do styling via JavaScript…

Screenshot 2023-07-14 at 09 32 10

### CODE REVIEW

File: src/stories/Dialog.stories.mdx [hunk 1]
```diff
@@ -241,7 +241,20 @@ An example with a long title that wraps onto the next line.
     name="listbox inside dialog"
     args={{
       title: '3D drawing style',
-      slots: `<template #body><AListbox model-value="" :options="[{ value: 0, label: 'Graphic' }, { value: 1, label: 'Rendered' }, { value: 2, label: 'Black and white' }]" /></template>`
+      slots: `<template #body><AListbox escape-overflow model-value="" :options="[{ value: 0, label: 'Graphic' }, { value: 1, label: 'Rendered' }, { value: 2, label: 'Black and white' }]" /></template>`
```
Comment (manuatarchilogic) [inline src/stories/Dialog.stories.mdx]: I tried to find a way to open the listbox for Chromatic UI snapshots, but failed. So it's as before (so not worse) but the fact that we don't open the listbox is what made us not realise that https://github.com/archilogic-com/ui-components/pull/330 introduced a regression...

File: src/components/Listbox.vue [hunk 2]
```diff
@@ -117,8 +127,9 @@ export default defineComponent({
 })
 </script>
 <template>
-  <Listbox v-model="model" as="div" class="group relative">
+  <Listbox v-model="model" as="div" class="group" :class="{ 'relative': !escapeOverflow }">
```
Comment (manuatarchilogic) [inline src/components/Listbox.vue:130]: by removing `relative`, the absolutely-positioned options panel does not position itself anymore relatively to the button. But that makes it lose the width constraints, which I had to re-add via JavaScript :(

**Review by luk-zap — APPROVED** [review]

File: src/components/Combobox.vue [hunk 3]
```diff
@@ -99,6 +99,13 @@ export default defineComponent({
     inline: {
       type: Boolean,
       default: false
+    },
+    /**
+     * allow the options panel to be rendered outside the flow of a container that has content that needs scrolling
+     */
+    escapeOverflow: {
+      type: Boolean,
+      default: false
```
Comment (romiekos) [inline src/components/Combobox.vue:108]: If PR introduces a new prop, I would expect to have at least one new story for Combobox and Listbox sections? Not only for Dialog.

Comment (manuatarchilogic) [reply src/components/Combobox.vue:108]: done

File: src/stories/Dialog.stories.mdx (hunk 1 above)
Comment (manuatarchilogic) [reply src/stories/Dialog.stories.mdx]: fixed, by not using a dialog...

File: src/stories/Combobox.mdx [hunk 4]
```diff
@@ -315,3 +315,17 @@ via `panelClasses` prop. Remember to add `overflow-auto` along with your height
 <Canvas>
   <Story id="components-combobox--panel-classes" />
 </Canvas>
+
+### Inside a container with overflow-auto
+
+Set the `escapeOverflow` prop to `true` to render the options panel outside the flow of the container.
+A typical usecase would be a Combobox inside a Dialog.
```
Comment (romiekos) [inline src/stories/Combobox.mdx]:
```suggestion
A typical use case would be a Combobox inside a Dialog.
```

File: src/stories/Listbox.mdx [hunk 5]
```diff
@@ -230,3 +230,16 @@ via `panelClasses` prop. Remember to add `overflow-auto` along with your height
 <Canvas>
   <Story id="components-listbox--panel-classes" />
 </Canvas>
+
+### Inside a container with overflow-auto
+
+Set the `escapeOverflow` prop to `true` to render the options panel outside the flow of the container.
+A typical usecase would be a Listbox inside a Dialog.
```
Comment (romiekos) [inline src/stories/Listbox.mdx]:
```suggestion
A typical use case would be a Listbox inside a Dialog.
```

File: src/stories/Listbox.mdx [hunk 6]
```diff
@@ -230,3 +230,16 @@ via `panelClasses` prop. Remember to add `overflow-auto` along with your height
 <Canvas>
   <Story id="components-listbox--panel-classes" />
 </Canvas>
+
+### Inside a container with overflow-auto
```
Comment (romiekos) [inline src/stories/Listbox.mdx:234]: Not sure why, I can't find this story in the listbox block https://649d652b7201581ed9f63642-wjosvtoqui.chromatic.com/?path=/docs/components-listbox--panel-classes. (expect to see it after the components-listbox--panel-classes).
image

Comment (manuatarchilogic) [reply src/stories/Listbox.mdx:234]: it's there, it's the last one. There's too much magic involved in the interplay between the `.mdx` file and the `.stories.ts` file

**Review by romiekos — APPROVED** [review]

File: src/components/Combobox.vue [hunk 7]
```diff
@@ -111,6 +118,9 @@ export default defineComponent({
 
     const inputRef = ref()
 
+    const comboboxButton = ref()
+    const optionsPanelWidth = computed(() => comboboxButton.value?.el.offsetWidth)
```
Comment (bnjm) [inline src/components/Combobox.vue:122]: it seems like `optionsPanelWidth` will always be the initially captured value of `el.offsetWidth`. Could this be an issue when the width changes, eg. resizable sidebars?

Comment (manuatarchilogic) [reply src/components/Combobox.vue:122]: maybe? How would you improve that?

Comment (bnjm) [reply src/components/Combobox.vue:122]: one approach would be a `ResizeObserver`? [vueuse](https://vueuse.org/core/useResizeObserver/) has a composable we could use or adapt. 

... however, I initially misread the PR as affecting the positioning of all combobox's. Given that it's additive with the `escapeOverflow` prop, this isn't a blocking issue at all

**Review by bnjm — APPROVED** [review]

**Pull request merged**